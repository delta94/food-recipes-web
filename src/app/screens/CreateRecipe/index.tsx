import React, { useState, useEffect, ChangeEvent } from 'react'
import { useDispatch } from 'react-redux'
import { Creators } from '../../../core/redux/store/ducks/recipe'
import BackupIcon from '@material-ui/icons/Backup'

import { Container, ContainerRecipe, BoxInput, BoxImage } from './styles'

const CreateRecipe = () => {
  const [selectedFile, setSelectedFile] = useState<File>()
  const [preview, setPreview] = useState<string>()

  const dispatch = useDispatch()

  const [name, setName] = useState('')
  const [ingredients, setIngredients] = useState('')
  const [mode_prepare, setMode_prepare] = useState('')
  const [time, setTime] = useState('')

  useEffect(() => {
    if (!selectedFile) {
      setPreview(undefined)
      return
    }

    const objectUrl = URL.createObjectURL(selectedFile)
    setPreview(objectUrl)

    return () => URL.revokeObjectURL(objectUrl)
  }, [selectedFile])

  const handleSelectFile = (e: ChangeEvent<HTMLInputElement>) => {
    if (!e.target.files || e.target.files.length === 0) {
      setSelectedFile(undefined)
      return
    }
    setSelectedFile(e.target.files[0])
  }

  function handleSubmit (event: ChangeEvent<HTMLFormElement>) {
    event.preventDefault()

    if (selectedFile) {
      dispatch(Creators.createRecipeRequest(selectedFile, name, ingredients, mode_prepare, time))
    }
  }

  return (
    <Container>
      <h1>Crie sua receita</h1>

      <ContainerRecipe>
        <form onSubmit={handleSubmit}>
          <div>
            <BoxImage preview={preview}>
              <label>
                <input type='file' name='image' onChange={handleSelectFile} placeholder="Insira sua imagem" />
                <BackupIcon fontSize='large' />
              </label>
            </BoxImage>

            <BoxInput>
              <label htmlFor='name'>Nome da receita:</label>
              <input
                name='name'
                placeholder='Insira o nome da sua receita'
                value={name}
                onChange={(e) => setName(e.target.value)}
              />
            </BoxInput>

            <BoxInput>
              <label htmlFor='ingredients'>Ingredientes:</label>
              <textarea
                name='ingredients'
                placeholder='Insira os ingredientes'
                value={ingredients}
                onChange={(e) => setIngredients(e.target.value)}
              />
            </BoxInput>

            <BoxInput>
              <label htmlFor='mode_prepare'>Modo de preparo:</label>
              <textarea
                name='mode_prepare'
                placeholder='Insira o modo de preparo'
                value={mode_prepare}
                onChange={(e) => setMode_prepare(e.target.value)}
              />
            </BoxInput>

            <BoxInput>
              <label htmlFor='time'>Tempo:</label>
              <input
                name='time'
                placeholder='Insira o tempo estimado'
                value={time}
                onChange={(e) => setTime(e.target.value)}
              />
            </BoxInput>

            <button type='submit'>Cadatrar receita</button>
          </div>
        </form>
      </ContainerRecipe>
    </Container>
  )
}

export default CreateRecipe
