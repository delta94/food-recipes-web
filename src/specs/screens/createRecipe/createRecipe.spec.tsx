import React from 'react'
import { fireEvent, getByText, render } from '@testing-library/react'
import { screen } from '@testing-library/dom'
import '@testing-library/jest-dom/extend-expect'

import CreateRecipe from '../../../app/screens/CreateRecipe'

import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'

const mockStore = configureStore([])

declare var global: any

describe('[SCREENS] createRecipe', () => {
  let store: any

  beforeEach(() => {
    store = mockStore({
      auth: {}
    })
  })

  it('Should render correctly', () => {
    const container = render(
      <BrowserRouter>
        <Provider store={store}>
          <CreateRecipe />
        </Provider>
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('Should return if onSubmit was called successfully', () => {
    const { getByText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <CreateRecipe />
        </Provider>
      </BrowserRouter>
    )

    expect(fireEvent.click(getByText('Cadatrar receita'))).toBeTruthy()
  })

  global.URL.createObjectURL = jest.fn()
  global.URL.revokeObjectURL = jest.fn()

  it('Should verify image is selected', () => {
    global.URL.createObjectURL = jest.fn(() => 'image/png')
    global.URL.revokeObjectURL = jest.fn(() => 'image/png')

    const { getByPlaceholderText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <CreateRecipe />
        </Provider>
      </BrowserRouter>
    )

    fireEvent.change(getByPlaceholderText('Insira sua imagem'), {
      target: {
        files: [new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png' })]
      }
    })

    URL.createObjectURL = jest.fn(() => 'image/png')

    expect(getByPlaceholderText('Insira sua imagem')).toBeTruthy()
  })

  it('Should return non selected file', () => {
    global.URL.createObjectURL = jest.fn(() => 'image/png')
    global.URL.revokeObjectURL = jest.fn(() => null)

    const { getByPlaceholderText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <CreateRecipe />
        </Provider>
      </BrowserRouter>
    )

    expect(fireEvent.change(getByPlaceholderText('Insira sua imagem'), {
      target: {
        files: null
      }
    })).toBeTruthy()
  })

  it('Should setSelectFile on send valid image', () => {
    global.URL.createObjectURL = jest.fn(() => 'image/png')
    global.URL.revokeObjectURL = jest.fn(() => null)

    const { getByPlaceholderText, getByText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <CreateRecipe />
        </Provider>
      </BrowserRouter>
    )

    fireEvent.change(getByPlaceholderText('Insira sua imagem'), {
      target: {
        files: [new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png' })]
      }
    })

    expect(fireEvent.click(getByText('Cadatrar receita'))).toBeTruthy()
  })
})
