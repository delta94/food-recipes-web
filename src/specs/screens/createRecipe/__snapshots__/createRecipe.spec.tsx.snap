// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SCREENS] createRecipe Should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="sc-bdnylx dvkmpa"
      >
        <h1>
          Crie sua receita
        </h1>
        <article
          class="sc-gtssRu eMVUtx"
        >
          <form>
            <div>
              <div
                class="sc-hKFyIo dnMOFA"
              >
                <label>
                  <input
                    name="image"
                    placeholder="Insira sua imagem"
                    type="file"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
                    />
                  </svg>
                </label>
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="name"
                >
                  Nome da receita:
                </label>
                <input
                  name="name"
                  placeholder="Insira o nome da sua receita"
                  value=""
                />
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="ingredients"
                >
                  Ingredientes:
                </label>
                <textarea
                  name="ingredients"
                  placeholder="Insira os ingredientes"
                />
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="mode_prepare"
                >
                  Modo de preparo:
                </label>
                <textarea
                  name="mode_prepare"
                  placeholder="Insira o modo de preparo"
                />
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="time"
                >
                  Tempo:
                </label>
                <input
                  name="time"
                  placeholder="Insira o tempo estimado"
                  value=""
                />
              </div>
              <button
                type="submit"
              >
                Cadatrar receita
              </button>
            </div>
          </form>
        </article>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="sc-bdnylx dvkmpa"
    >
      <h1>
        Crie sua receita
      </h1>
      <article
        class="sc-gtssRu eMVUtx"
      >
        <form>
          <div>
            <div
              class="sc-hKFyIo dnMOFA"
            >
              <label>
                <input
                  name="image"
                  placeholder="Insira sua imagem"
                  type="file"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
                  />
                </svg>
              </label>
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="name"
              >
                Nome da receita:
              </label>
              <input
                name="name"
                placeholder="Insira o nome da sua receita"
                value=""
              />
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="ingredients"
              >
                Ingredientes:
              </label>
              <textarea
                name="ingredients"
                placeholder="Insira os ingredientes"
              />
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="mode_prepare"
              >
                Modo de preparo:
              </label>
              <textarea
                name="mode_prepare"
                placeholder="Insira o modo de preparo"
              />
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="time"
              >
                Tempo:
              </label>
              <input
                name="time"
                placeholder="Insira o tempo estimado"
                value=""
              />
            </div>
            <button
              type="submit"
            >
              Cadatrar receita
            </button>
          </div>
        </form>
      </article>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
